File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ret = 0
        def getDiameter(node):
            nonlocal ret 
            if not node:
                return 0
            l, r = getDiameter(node.left), getDiameter(node.right)
            ret = max(ret, l+r)
            return 1 + max(l, r)
        getDiameter(root)

        return ret

File: Data Structures & Algorithms/partition-labels/submission-1.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        vector<int> map(26, -1);
        for(int i = 0; i < s.size(); i++){
            char ch = s[i];
            map[ch - 'a'] = i; 
        }

        vector<int> res;
        int left = 0, right = 0;
        int last = -1;
        while(right < s.size()){
            last = max(last, map[s[right]-'a']);
            // cout << s[right]-'a' << " " << map[s[right]-'a'] << " " << last << endl;
            if(last == right){
                res.push_back(right - left + 1);
                left = right + 1;
            }
            right++;
        }
        return res;
    }
};
